### PR TITLE
docs(chart_tempalte_primer): add template primer

### PR DESCRIPTION
**Pull request by technosophos**
This adds a detailed document on writing templates in Helm. This is NOT
a best-practices guide, but a thorough introduction to templates in
Helm.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/helm/1500)

### CODE REVIEW

File: docs/chart_template_guide/debugging.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Debugging Templates
+
+Debugging templates can be tricky simply because the templates are rendered on the Tiller server, not the Helm client. And then the rendered templates are sent to the Kubernetes API server, which may reject the YAML files for reasons other than formatting.
+
+There are a few commands that can help you debug.
+
+- `helm lint` is your go-to tool for verifying that your chart follows best practices
+- `helm install --dry-run --debug`: We've seen this trick already. It's a great way to have the server render your templates, then return the resulting manifest file.
```
Comment (prydonius) [inline docs/chart_template_guide/debugging.md]: Something I learned the other day, to solve the issue of YAML formatting issues when Helm tries to parse (which, if it fails to do, it won't output anything), you can comment out the template and run `--dry-run --debug`.

**Comment by prydonius** [comment]
Had a brief look and this is really really great, thanks so much for putting this together @technosophos!

**Comment by gtaylor** [comment]
I almost hate to even bring something so petty up, but do we have any agreement on whether to include spaces before and after the templated curly braces? These examples are likely to be referenced as starting points, which is the only reason I bring this up. It seems like many/most of the charts pad the curly braces. I see a ton of examples in Go land that don't pad.

Is this something we want to decide on either way for the sake of consistency?

File: docs/chart_template_guide/debugging.md (hunk 1 above)
Comment (technosophos) [inline docs/chart_template_guide/debugging.md]: Ooo... good idea

**Comment by technosophos** [comment]
@gtaylor That is an excellent point, actually. The padding is probably a Good Idea (TM) just because of the possibility of screwing things up with the `{{-` and `-}}` if you don't pad. I'll take a pass through for consistency.

**Pull request merged**